jaeger-spark-dependencies-common/src/main/java/io/jaegertracing/spark/dependencies/SpansToDependencyLinks.java

```diff
@@ -35,26 +38,102 @@
+            if (parents != null) {
+                if (parents.size() > 1) {
+                    // multiple possible parents so parent is zipkin shared span
+                    // therefore we choose the one with the greatest start time (e.g. server span)
+                    // we cannot choose client span because it's not a true parent!
+                    Span sharedParent = null;
+                    for (Span sharedSpan: parents) {
+                        if (sharedParent == null || sharedSpan.getStartTime() > sharedParent.getStartTime()) {
+                            sharedParent = sharedSpan;
+                        }
+                    }
+                    result.add(new Dependency(sharedParent.getProcess().getServiceName(), span.getProcess().getServiceName()));
+                } else {
+                    // this is jaeger span or zipkin native (not shared!)
+                    Span parent = parents.iterator().next();
+                    if (parent.getProcess() == null || parent.getProcess().getServiceName() == null) {
+                        continue;
+                    }
+                    result.add(new Dependency(parent.getProcess().getServiceName(), span.getProcess().getServiceName()));
+                }
+            }
         }
         return result;
     }
+
+    /**
+     * @param span span
+     * @param spanMap span map, key is spanId and value set of spans
+     * @return true if there is a span in the set which started before given span.
+     */
+    private boolean existsSpanWhichStartedBefore(Span span, Map<Long, Set<Span>> spanMap) {
+        Set<Span> sharedSpans = spanMap.get(span.getSpanId());
+        if (sharedSpans != null && sharedSpans.size() > 1) {
+            for (Span other: sharedSpans) {
+                if (span.getStartTime() > other.getStartTime()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private List<Dependency> sharedSpanDependencies(Map<Long, Set<Span>> spanMap) {
+        List<Dependency> dependencies = new ArrayList<>();
+        // create links between shared spans
+        spanMap.values().forEach(spans -> {
+            List<Span> spanList = new ArrayList<>(spans);
+            for (int i = 0; i < spanList.size(); i++) {
+                for (int j = i + 1; j < spanList.size(); j++) {
+                    createLinkSharedSpan(spanList.get(i), spanList.get(j))
+                        .ifPresent(dependency -> dependencies.add(dependency));
+                }
+            }
+        });
+
+        return dependencies;
+    }
+
+    private Optional<Dependency> createLinkSharedSpan(Span spanA, Span spanB) {
+        // return empty if the span is not shared
+        if (!spanA.getSpanId().equals(spanB.getSpanId())) {
+            return Optional.empty();
+        }
+
+        Dependency dependency;
+        if (spanA.getStartTime() < spanB.getStartTime()) {
```

Comment:
The way we did it for Zipkin model was to search for same-ID spans that represent server and client. When Jaeger receives Zipkin model it converts cs/cr/ss/sr into `span.kind` tags, so you can determine the direction of the link from span.kind=client => span.kind=server.

The timestamp comparison only makes sense if you're comparing spans from the same process.